kvdb/src/lib.rs

```diff
@@ -134,21 +115,22 @@ impl DBTransaction {
 ///
 /// The API laid out here, along with the `Sync` bound implies interior synchronization for
 /// implementation.
-pub trait KeyValueDB: Sync + Send {
+pub trait KeyValueDB<K: AsRef<[u8]>, V>: Sync + Send {
```

Comment:
I am having doubts about how this is actually works out in consuming code. As it stands before this PR, consuming code can use different key types freely. For example in `parity-ethereum` we use `[u8; 12]`, `H256` and `&[u8]` (and possibly others) without any particular gymnastics required. With this PR and `K` set to be `Vec<u8>` all kinds of "fun" conversions become necessary for little apparent gain.

I know this is a lot to ask @tafia but could you take a look at for example `journaldb` ([here](https://github.com/paritytech/parity-ethereum/tree/master/util/journaldb)) and show me how you'd convert it to use this PR? When I try it gets painful fast.